web/src/main/webapp/my-app/layout/controllers.js

```diff
@@ -251,9 +256,17 @@ define(['angular', 'jquery'], function(angular, $) {
           if (response === 'undo') {
             // Add the removed widget back to the layout
             $scope.layout.splice(data.removedIndex, 0, data.removedWidget);
+
+            // Delete the last fname added to the removal array
+            $scope.widgetsToRemove.pop();
           } else {
-            // Persist changes
-            saveLayoutRemoval(data.removedWidget.fname, data.removedIndex);
+            // Save deletion of any widgets in the tracking array
+            for (var i = 0; i < $scope.widgetsToRemove.length; i++) {
+              var fname = $scope.widgetsToRemove[i];
+              saveLayoutRemoval(fname);
+              // Remove saved deletions from tracking array
+              $scope.widgetsToRemove.splice(i, 1);
```

Comment:
um. Does this `splice` change the state of the `widgetsToRemove` array while this loop is iterating through it? Does indexing into the array with `i` remain valid across this splicing?